### PR TITLE
Feature/poi 4.1.0

### PR DESCRIPTION
**Pull request by spoltier**
Upgrade to POI 4.1.0
* Java 8 is required
* Use java.time instead of the Joda library, small changes in parsing / formatting dates and times
* Adapt to small changes in API
* Update dependencies in the pom.xml

### CODE REVIEW

File: src/main/java/com/miraisolutions/xlconnect/Workbook.java [hunk 1]
```diff
@@ -458,11 +459,11 @@ private void writeData(DataFrame data, Sheet sheet, int startRow, int startCol,
                             setMissing(cell);
                         else {
                             if(Double.isInfinite(doubleValues[j])) {
-                              cell.setCellType(CellType.ERROR);
```
Comment (martinstuder) [inline src/main/java/com/miraisolutions/xlconnect/Workbook.java]: Is this now done implicitly?

File: src/main/java/com/miraisolutions/xlconnect/Workbook.java [hunk 2]
```diff
@@ -458,11 +459,11 @@ private void writeData(DataFrame data, Sheet sheet, int startRow, int startCol,
...
+
                               cell.setCellErrorValue(FormulaError.NA.getCode());
                             } else {
                               cell.setCellValue(doubleValues[j]);
-                              cell.setCellType(CellType.NUMERIC);
```
Comment (martinstuder) [inline src/main/java/com/miraisolutions/xlconnect/Workbook.java]: Same here. Is the cell type set implicitly by serCellValue?

File: src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java [hunk 3]
```diff
@@ -110,10 +115,10 @@
                       n = Integer.parseInt(Character.toString(format.charAt(++i)));
                   }
                   
-                  builder.appendSecondOfMinute(2);
+                  builder.appendValue(ChronoField.SECOND_OF_MINUTE);
                   if(n > 0) {
                       builder.appendLiteral('.');
-                      builder.appendFractionOfSecond(n, n);
+                      builder.appendFraction(ChronoField.SECOND_OF_MINUTE,2,2, true);
```
Comment (martinstuder) [inline src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java]: Shouldn't this be using n here?

File: src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java [hunk 4]
```diff
@@ -153,24 +158,21 @@
               // the 2004 POSIX standard, but it does also say �it is expected
               // that in a future version the default century inferred from a
               // 2-digit year will change�.
-              builder.appendTwoDigitYear(1968, true);
+              builder.appendValueReduced(ChronoField.YEAR, 2, 2, 1969);
```
Comment (martinstuder) [inline src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java]: Is the 1969 insteadof 1968 intended?

File: src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java [hunk 5]
```diff
@@ -179,19 +181,27 @@
...
     public String format(Date d, String format) {
         StringBuffer sb = new StringBuffer();
-        getFormatter(format).printTo(sb, new org.joda.time.DateTime(d));
+        getFormatter(format).formatTo(ZonedDateTime.ofInstant(d.toInstant(), ZoneId.systemDefault()), sb);
         return sb.toString();
     }
 
     public Date parse(String s, String format) {
-        return getFormatter(format).parseDateTime(s).toDate();
+        java.time.format.DateTimeFormatter formatter = getFormatter(format);
```
Comment (martinstuder) [inline src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java]: Is this according to the logic Joda's parseDateTime used? java.util.Date doesn't have any timezone information attached to it AFAIK.

File: src/main/java/com/miraisolutions/xlconnect/Workbook.java (hunk 2 above)
Comment (spoltier) [reply src/main/java/com/miraisolutions/xlconnect/Workbook.java]: Correct, see https://poi.apache.org/apidocs/4.1/org/apache/poi/ss/usermodel/Cell.html#setCellType-org.apache.poi.ss.usermodel.CellType-

File: src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java (hunk 4 above)
Comment (spoltier) [reply src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java]: Yes, the range implied by the base year is inclusive, this gives the range 1969-2068 as specified by the comment (unchanged). 1968 would imply the range 1968-2067.

File: src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java (hunk 5 above)
Comment (spoltier) [reply src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java]: @martinstuder I changed the code to use LocalDateTime. It still requires a default zone offset / id to convert to Date via Instant.

File: src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java (hunk 3 above)
Comment (spoltier) [reply src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java]: Done, thank you :smiley:

File: src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java (hunk 5 above)
Comment (spoltier) [reply src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java]: update: I tested that the (correct) behavior described in the [comments](https://github.com/miraisolutions/xlconnect/issues/50#issuecomment-460316521) of issue 50 remains unchanged.

File: src/main/java/com/miraisolutions/xlconnect/Workbook.java (hunk 1 above)
Comment (spoltier) [reply src/main/java/com/miraisolutions/xlconnect/Workbook.java]: same as below

File: src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java (hunk 4 above)
Comment (spoltier) [reply src/main/java/com/miraisolutions/xlconnect/utils/RPOSIXDateTimeFormatter.java]: Checked that the [R documentation|https://stat.ethz.ch/R-manual/R-devel/library/base/html/strptime.html] as of this comment, still describes the same interval (and R 3.6.1 also parses dates according to it).

**Pull request merged**